accepted/better-obsoletion/better-obsoletion.md

```diff
@@ -0,0 +1,345 @@
+  something doesn't work. Only few developers research guidance and
+  best-practice proactively. But even if they did: guidance changes over time
+  and some technologies that looked like good ideas turn out to be unviable. We
+  need a mechanism that informs developers after the fact.
+
+Obsoletion via the `[Obsolete]` attribute works pretty well when: only a few
+APIs are affected, it's easy to remove their usage, and guidance can be provided
+as a short one-line string.
+
+In practice these requirements are often not satisfiable:
+
+* **Technologies have large API surfaces**. Obsoleting an entire technology
+  means that we'll have to annotate all entry points to that technology, which
+  often means obsoleting a *set of APIs*, such as types and static members.
+
+* **Removing obsoleted APIs isn't always practical**. It would be nice if we
+  could always remove dependencies on obsoleted technologies but this isn't
+  always possible if they are used heavily in existing code bases. In those
+  cases, you typically want a way to ignore obsoletion diagnostics regarding
+  specific technologies but not others so that you don't accidentally start
+  depending on obsoleted technologies when you're evolving your code base.
+
+* **Guidance is complex**. Obsoleted APIs rarely have a drop-in replacement.
+  After all, we don't obsolete based on their naming or shape but usually due to
+  fundamental design issues. That means the replacements are typically a nuanced
+  choice and require at least a page of explanation, rather than a one-liner. 
+
+This document proposes changes to how APIs can be obsoleted in a more structured
+way.
+
+## Scenarios and User Experience
+
+### Targeted Suppression
+
+Abigail is a developer and maintains the business logic layer component for
+Fabrikam Billing Services. It was written in 2004 and makes heavy use of
+non-generic collections by inheriting from them. Refactoring the code base to
+use generic collections would be possible but a ton of work of dubious value as
+the code is well-tested. Last year, the code was ported to .NET Core 2.1.
+
+The component is designed for a traditional two-tier environment and includes
+data access that handles retrieving and storing business objects from and to SQL
+Server. Abigail is now tasked to refactor that code to use REST-based services
+instead so that the application can eventually be converted to a cloud-based
+offering.
+
+Abigail would like to use the brand new JSON APIs so she updates the project
+from .NET Core 2.1 to .NET Core 3.1. However, after upgrading the notices that
+we obsoleted non-generic collections and she gets a bunch of warnings against
+her code:
+
+| Code    | Description              | Project                        |
+|---------|--------------------------|--------------------------------|
+| BCL0006 | 'ArrayList' is obsolete  | Fabrikam.Billing.BusinessLogic |
+| BCL0006 | 'Hashtable' is obsolete  | Fabrikam.Billing.BusinessLogic |
```

Comment:
FWIW, Hashtable is a perfectly good type that does not have direct equivalent anywhere else in the framework when it comes to multi-threading. We should stop trying to obsolete it. It does not work. Replacing it with something else just creates bugs.

`BinaryFormatter` is a good example of a type to obsolete.

Reply:
Sure, we can discuss that when we actually apply the attribute. I've no strong opinions on `Hashtable`. But duly noted 😄